shark_turbine/kernel/wave/utils.py

```diff
@@ -378,3 +389,132 @@ def erase_graph(graph: fx.Graph):
         for user in node.users:
             graph.erase_node(user)
         graph.erase_node(node)
+
+
+def get_users(
+    node: fx.Node, reduction: fx.Node = None
+) -> tuple[list[fx.Node], fx.Node]:
+    """
+    Return the users of a node, propagating through reductions.
+    """
+    users = []
+    for user in node.users:
+        custom = get_custom(user)
+        if isinstance(custom, Reduction):
+            # Map init arg to iter arg
+            reduction = custom
+            init_arg_idx = custom.init_args.index(node)
+            users.append(custom.iter_args[init_arg_idx])
+            continue
+        if isinstance(custom, Output) and reduction:
+            # Map output to get result
+            return_vals = custom.return_vals[0]
+            get_results = sorted(
+                [x for x in reduction.users if isinstance(get_custom(x), GetResult)],
+                lambda x: get_custom(x).res_idx,
+            )
+            if isinstance(return_vals, list):
+                output_idx = return_vals.index(node)
+                users.append(get_results[output_idx])
+            else:
+                users.append(get_results[0])
+            continue
+        users.append(user)
+    return users, reduction
+
+
+def get_inputs(
+    node: fx.Node, reduction: fx.Node = None
+) -> tuple[list[fx.Node], fx.Node]:
+    """
+    Return the inputs of a node, propagating through reductions.
+    """
+    inputs = []
+    for input in node.all_input_nodes:
+        custom = get_custom(input)
+        if isinstance(custom, GetResult):
+            reduction = custom.value
+            assert isinstance(
+                reduction, Reduction
+            ), "GetResult must be used by a Reduction"
+            # Map get result to output
+            inputs.append(reduction.outputs[custom.res_idx])
+            continue
+        if isinstance(custom, IterArg):
+            # Map iter args to init args
+            iter_arg_idx = reduction.iter_args.index(node)
+            inputs.append(reduction.init_args[iter_arg_idx])
+            continue
+        inputs.append(input)
+    return inputs, reduction
+
+
+def bfs(
+    node: fx.Node,
+    visited: list[fx.Node],
```

Comment:
I don't think we need the `visited`/`queue` right? especially since it's iterative search?

Reply:
So the idea here is that we want to find all the transitive dependencies of the lhs, rhs (backward slice) and acc (forward slice). So find all the nodes that could affect lhs and rhs and all the nodes that are affected by acc. visited keeps track of all these nodes. queue just helps with the bfs. The reason we do that is so that now we are assigning global per dimension indices. So we need to "specialize" these indices by identifying whether they lie in the backward slices of the lhs, rhs or forward slice of the acc. When we have multiple mma ops, then we will have some of these slices intersecting. Specifically, the forward acc slice will intersect with the backward slices of the lhs and rhs. This would be the equivalent of a conflict.

Reply:
That makes sense, although what I meant was couldn't we initialize `visited`/`queue` inside the function? Since we always send empty list for both visited and queue anyways?

Reply:
Ah yes, that makes sense. Will change.